### DISCUSSION TITLE
Retrieving results from CI tool

### DISCUSSION THREAD

**Discussion by shaqb** [ludeknovy/jtl-reporter#207] [Q&A]
Hi team! We're just exploring the concept of shift left testing and we came across your tool for sending our results data to as it gives us lovely rich dashboards!

One thing we want to do is pass or fail a build based on whether a performance test has deviated from previous runs (say 10% deviation on the 95th percentile response time for a particular API). 

I can see JTL reporter can send alerts based on deviation but is there a way to call an API to get the same status which we can call at the end of our run as a pipeline stuff, and pass/fail the build accordingly?

Our pipeline would look something like:

Developer merges code -> build/deploy -> run performance test -> push to JTL Reporter -> check against baseline data -> pass/fail.

Thanks!

Shaq

**Comment by ludeknovy** [comment]
Hi @shaqb !
If I understood correctly your query - you would like to send the report status from JtlReporter to CI? And based on the response adjust the build status accordingly?

**Comment by shaqb** [comment]
Hi @ludeknovy - yes that's correct. Or alternatively, the CI can query the JTLReporter at the end of the build and the response from the JTLReporter will determine whether the build passes or fails. This is so we can stop builds from being progressed in our CI flow.

**Reply by ludeknovy** [reply]
The problem here is, that the processing is asynchronous. Doing it from CI should be possible, and you will need to create a script that will keep polling until the report is fully processed. If you upload the report API returns the `itemId` (report id), which you can use for polling. This way you can get any data that's available to you in the report detail.

**Reply by shaqb** [reply]
Ok that should be simple enough to do from our CI tool. I can post, grab the ID and then poll until I get the response I need.

Do you have API I need to poll and what data comes back in it?

**Reply by ludeknovy** [reply]
The endpoint you want to poll is `/api/projects/<projectName>/scenarios/<scenarioName>/items/<id>`. Be warned that the response might be large.

In the root of the returned json object, you can find `thresholds`. If threshold feature is not enabled, `null` is returned. If enabled and it passes, the following is returned: 
```
{ thresholds: { passed: true }
```

A warning, though, in your initial message you mentioned a deviation in response time for a particular API. This is not how it currently works - it calculates the deviation from aggregated metrics. From here you can investigate it further: https://github.com/ludeknovy/jtl-reporter-be/blob/9df3a6e54a78c8d0217710bfc9948d46056fb765/src/server/controllers/item/shared/item-data-processing.ts#L126

**Reply by shaqb** [reply]
Thanks @ludeknovy - will look to implement and see it in action. Thanks for the great work on this project btw! It's very impressive

**Reply by ludeknovy** [reply]
Glad you like it! 

Actually, I think I will change the behavior of the thresholds the way you mentioned - it will compare the values to a baseline report, and it will do it for all the labels.

**Reply by shaqb** [reply]
That would be awesome!

**Reply by shaqb** [reply]
There is a performance plugin for Jenkins which does similar: http://jenkinsci.github.io/performance-plugin/Reporting.html

But:

- It's not CI agnostic (only works in Jenkins)
- The graphs are terrible
- Retention is not great

But I like the idea of passing/failing builds based on a deviation from a relative baseline or an absolute threshold.  Also being able to fail builds if too high an error rate was seen.

**Reply by ludeknovy** [reply]
WIP 👁️ 

<img width="1423" alt="Screenshot 2023-04-02 at 21 55 33" src="https://user-images.githubusercontent.com/13610612/229375972-cfafe665-770d-4468-ac53-15958091a175.png">

<img width="1406" alt="Screenshot 2023-04-02 at 21 58 18" src="https://user-images.githubusercontent.com/13610612/229376021-814ccace-7835-40c7-9441-709a5e8dab60.png">

**Reply by ludeknovy** [reply]
@shaqb what would be the desired behavior in case a baseline report is deleted or changed?

**Reply by ludeknovy** [reply]
@shaqb Let me re-phrase my question - what should happen for the current reports? 
Case 1: the baseline report is deleted - for the reports they contain regression alert, I was thinking of removing it. Of course for newly uploaded reports, it won't get calculated.
Case2: the baseline report is updated - I was thinking about updating the regression alerts asynchronously.

**Comment by shaqb** [comment]
That’s awesome!

**Comment by shaqb** [comment]
I think in that situation we can’t report against and maybe have a
different return code or status?

**Comment by shaqb** [comment]
For case 2 would you look to recalculate on an interval or at point of
request for the baseline and we would continue to poll until it’s ready?
What would the sequence look like?

**Reply by ludeknovy** [reply]
This would get recalculated after selecting another report as a baseline (it's done through the report detail update). 
It would have no impact on your flow (it could, but the changes are very little actually, as the baseline report change would need to happen during the new report processing). The case 2 is much more about whether the reports should be re-calculated for the past reports.

**Reply by ludeknovy** [reply]
https://github.com/ludeknovy/jtl-reporter-be/pull/213
https://github.com/ludeknovy/jtl-reporter-fe/pull/316

**Comment by shaqb** [comment]
Ok that makes sense.

**Comment by shaqb** [comment]
You’ve been busy!

**Reply by ludeknovy** [reply]
released in v4.8.2
https://github.com/ludeknovy/jtl-reporter/blob/main/docker-compose.yml

**Comment by shaqb** [comment]
Hi @ludeknovy - I'm now in the process of getting this integrated. Is there any documentation on how to retrieve the data, and also how to set thresholds for the baseline comparison (% or absolute deviation, number of runs to base it on, etc)

**Reply by ludeknovy** [reply]
To retrieve the data, you will need an [API token](https://jtlreporter.site/docs/guides/administration/api-token), and with that you will be able to call `/api/projects/<project_name>/scenarios/<scenario_name>/items/<id>`

To set up the thresholds go to Project->scenario->settings, and please keep in mind that you need to have the baseline set up.
You can do so from the report settings modal (accessible via report detail or test runs listing) and checking the "Set as base run" checkbox.

![Screenshot 2023-05-05 at 12 38 15](https://user-images.githubusercontent.com/13610612/236436692-b090fc3d-958e-4994-b0f1-ba6e02632b57.png)
![Screenshot 2023-05-05 at 12 39 50](https://user-images.githubusercontent.com/13610612/236437020-40d8699b-0fac-4208-a623-c26a8cd76c2b.png)

**Comment by shaqb** [comment]
Ah ok got it. Is it possible for the "base run" data to be automatically calculated based on the last x runs where x is a configurable number? 

So as an example I want to see if my latest run has deviated by more than 10% of the last 10 runs. This way the test reporting is completely dynamic and we don't need to regularly go in and manually update the latest base run (which we would need to do if performance got better or worse in a given release).

**Reply by ludeknovy** [reply]
No, this is not currently possible. Actually, it was working similar to this in the past. But you can set the baseline just once. There is a disadvantage from the approach you are describing; over a long period of time, the performance could really degrade without noticing/ triggering the threshold. So I think it really makes much more sense to compare it to a baseline report in this case.

**Reply by shaqb** [reply]
Fair point. Anyway let me play around with it and see what the output looks like. I still can't believe you've made such a comprehensive tool like this!

**Reply by ludeknovy** [reply]
Keen to hear your feedback!

**Reply by shaqb** [reply]
Hey @ludeknovy - been playing around with this. Is it the threshold block I need to look at?

```
"thresholds": {
    "passed": false,
    "results": [
      {
        "label": "RQ-BrowsePDP",
        "passed": true,
        "result": {
          "errorRate": {
            "passed": true,
            "diffValue": 100
          },
          "percentile": {
            "passed": true,
            "diffValue": 57.72285966460724
          },
          "throughput": {
            "passed": true,
            "diffValue": null
          }
        }
      },
      {
        "label": "RQ-InventoryAvailability",
        "passed": true,
        "result": {
          "errorRate": {
            "passed": true,
            "diffValue": 100
          },
          "percentile": {
            "passed": true,
            "diffValue": 24.846446329336064
          },
          "throughput": {
            "passed": true,
            "diffValue": null
          }
        }
      },
      {
        "label": "RQ-ProductList",
        "passed": false,
        "result": {
          "errorRate": {
            "passed": true,
            "diffValue": 100
          },
          "percentile": {
            "passed": false,
            "diffValue": 107.27272727272728
          },
          "throughput": {
            "passed": true,
            "diffValue": null
          }
        }
      }
    ],
    "thresholds": {
      "errorRate": "5",
      "percentile": "5",
      "throughput": "5"
    }
  },

```

What's your view on having a separate API endpoint for this just to make it easier to parse for our CI tool...

**Reply by ludeknovy** [reply]
Hi @shaqb yes, that's it. How would a separate endpoint make the parsing easier? Honestly, I don't see any benefit here.

**Comment by shaqb** [comment]
Hi Ludeknovy. Poor choice of words on my part. What I meant was to have an
endpoint specifically just for the pass/fail status. But I can see that is
quite a specific use case and the information is already there.

**Reply by ludeknovy** [reply]
Yes, from my point of view, it's redundant.

**Comment by shaqb** [comment]
Hey @ludeknovy so we've started integrating the JTLReporter into our CI pipelines and I think we might have an issue. When I submitted the result I got an itemId back. However, this test run is still processing so when I call the API to get the status I get back the following response:

`{
  "message": "Not Found"
}`

Is it possible to change the output to say 'processing' or something like that? I can then poll with the CI tool until the actual json of the test run is returned.

I can see the processing here: 
![image](https://github.com/ludeknovy/jtl-reporter/assets/19668009/6e41f4f8-5dd0-452d-9e47-a137f2ae5436)

But the API is showing the Not Found above.

**Reply by ludeknovy** [reply]
Hi @shaqb! Changing that API is not as easy as it sounds. You can achieve the desired behavior by using following endpoint:
`/api/projects/<project_name>/scenarios/<scenario_name>/processing-items`
in the response it returns:

```
{
    "failed": [],
    "inprogress": [
        {
            "id": "3215c173-ebc1-469a-b92f-5eddadc53dde",
            "reportStatus": "in_progress",
            "uploadTime": "2023-04-03T16:58:29.035Z"
        }
    ]
}
```

And once it's not there anymore, it means the processing finished.

**Reply by shaqb** [reply]
Nice! That will do the trick. I'll add that as a step in our pipeline.
